File: 2021/10/10/BinaryGap.java
/**
 * https://leetcode.com/problems/binary-gap/
 * @author Andrew Jarombek
 * @since 10/10/2021
 */

public class BinaryGap {
    public int binaryGap(int n) {
        if (n < 2) {
            return 0;
        }

        int gap = 0;
        boolean oneFound = false;
        int lastOne = 0;
        int i = 0;

        while (n > 0) {
            if (n % 2 == 1) {
                if (oneFound) {
                    if (i - lastOne > gap) {
                        gap = i - lastOne;
                    }
                } else {
                    oneFound = true;
                }

                lastOne = i;
            }

            i++;
            n /= 2;
        }

        return gap;
    }
}

File: 2023/05/05/binaryGap.js
/**
 * https://leetcode.com/problems/binary-gap/
 * @author Andrew Jarombek
 * @since 5/5/2023
 */

const binaryGap = function(n) {
    let result = 0;
    let index = 0;
    let prev = null;

    while (n > 0) {
        if (n % 2 === 1) {
            if (prev !== null) {
                result = Math.max(result, index - prev);
            }

            prev = index;
        }

        index++;
        n >>= 1;
    }

    return result;
}

console.assert(binaryGap(22) === 2);
console.assert(binaryGap(8) === 0);
console.assert(binaryGap(5) === 2);
